### ISSUE TITLE
Authorization error at consumption of the on-premise odata service (cloud-sdk for js)

### ISSUE THREAD

**Issue by EvgObraztsov** [SAP/cloud-sdk#171]
**Describe the bug**
Authorization error at consumption of the on-premise odata service in a cap application deployed to the SAP CF live tenant that uses the custom odata client library generated by the cloud sdk generator via the "DestinationNameAndJwt" interface.

**To Reproduce**
1. Set up: 
- a cap app to read data from the external on-premise odata service;
- approuter
- express.js in the srv module - to pass a user jwt to a handler of a sevice provider in the cap app
- odata client library generated by the cloud sdk generator
- uaa service
- destination service
- connectivity service
- destination to the on-premise system with the Basic authentication
- cloud connector in the CF env to the SAP internal on-premise system
2. Execute
const osEntity = await sdkEntity.requestBuilder().getAll().execute({ destinationName: destinationName, jwt: userToken }) 

3. Confirm ...
![DestinationNameAndJwt_debug_1](https://user-images.githubusercontent.com/32936758/83625714-12ad0980-a59d-11ea-9359-70d9eae1b564.png)
![DestinationNameAndJwt_debug_3](https://user-images.githubusercontent.com/32936758/83625725-16d92700-a59d-11ea-82d6-f8fef01e8b2c.png)
![DestinationNameAndJwt_debug_2](https://user-images.githubusercontent.com/32936758/83625726-16d92700-a59d-11ea-9116-d332389bb0df.png)
![DestinationNameAndJwt_debug_1](https://user-images.githubusercontent.com/32936758/83625779-2a848d80-a59d-11ea-8fd2-bb3f67e9e6e7.png)

Error: Request failed with status code 401

**Expected behavior**
To get data from the on-premise system via the exposed by a service provider link in a cap app 

**Screenshots**

**Used Versions:**
 - node version via `node -v` - I don't know the node version used in SAP SCP CF
 - npm version via `npm -v`- I don't know the npm version used in SAP SCP CF
 - SAP Cloud SDK version you used as dependency - "@sap-cloud-sdk/core": "^1.20.1"
**Code Examples**
__________________________________________________________
```
const cds = require('@sap/cds')

function getUserJWT (req) {
	if(typeof(req._) !== 'undefined') {
		return retrieveJwt(req._.req)
	} else {
		return ""
	}
}
module.exports = async function (){

    const { Entity: sdkEntity } = require('./odata-client/odata-service')
	const destinationName = 'dest_name'
	
	this.on('READ', 'sdkEntity', async (req) => {

		var userToken = getUserJWT(req)	
		
		try{
			const osEntity = await sdkEntity
			.requestBuilder()
			.getAll()
			.execute({ destinationName: destinationName, jwt: userToken}) // [!Error]([url](url)) is fired
			
			let mappedEntity = await osEntity.map(bp => serializeEntity(bp, 
                         sdkEntity))
			
			return mappedEntity
		} catch (e) {
			debugger;
			console.error("Error: " + e.message)
			console.log("Stack: " + e.stack)
		}
	})
```
________________________________________________________________________
**Log file**
If applicable, add your log file or related error message. Again, please remove your sensitive information.

**Additional context**

**Please check the following points**
- [ ] Have you provided detailed description about your issue?
- [ ] Have you shared all the necessary information to help the cloud-sdk developer reproduce the issue?
- [ ] Have you provided the expected behaviour?
- [ ] Have you attached some useful screenshots?
- [ ] Have you provided the node/npm/cloud-sdk versions?
- [ ] Have you shared your code snippets as a minial example?
- [ ] Have you provided you log file or related error messages?
- [ ] Have you searched for related issues on [Stackoverflow](https://stackoverflow.com/questions/tagged/sap-cloud-sdk) or [answers.sap.com](https://answers.sap.com/tags/73555000100800000895)?

**Comment by marikaner** [comment]
Hello @EvgObraztsov,

thank you for reporting this. Can you give us some details on what `retrieveJwt(req._.req)` returns? Is it `undefined`, an empty string or something meaningful?

**Comment by EvgObraztsov** [comment]
Hi @marikaner,
It's a user jwt.

**Comment by EvgObraztsov** [comment]
P.S. By the way the cloud connector is configured properly and I’m able to consume the odata service from the on-premise system in an ordinary mta app by reading environment variables and sending a http-request to the destination with configured connectivity-proxy settings.

**Comment by FrankEssenberger** [comment]
Hello EvgObraztsov,

I thought about the issue a bit and checked also our E2E tests which are also using the connectivity service and the cloud connector and they are green. So there is not something conceptional wrong but somewhere in the setup there must be a problem. 

In order to limit the sources, I would like to see if it is already the S/4 system sending the 401 or the connectivity proxy. So could you go to transaction `SMICM` and set the trace level via the GOTO button to 3:

![image](https://user-images.githubusercontent.com/56544999/84162820-1941f200-aa71-11ea-9ea0-680fc3d8c37f.png)

Then trigger the request. Afterwards use transaction `ST11` to investigate the `dev_icm` file.If the file is very large consider also to reset the file in `SMICM`. Then the investigation is also easier. I recommend to download the file to the disk, because the view in the GUI is very cumbersome to use. Look for lines like this:

```
[Thr 140528442128128]   HTTP/1.1 401 Unauthorized
```

If they are there we know that the request reach already the S/4 system. Since the setup with the log files is a bit difficult please also make a positive test that you see a request you triggered in the logs so that you are sure the a missing `401` in the log really means no request reached the system.

**Comment by EvgObraztsov** [comment]
@FrankEssenberger 
Hi Frank! 
Thank you for your input!
It will take me some time to check it with the person who has such kind of access to the on-premise system. Once we figure it out I’ll get back to you.
Best Regards,
Evgeniy

**Comment by EvgObraztsov** [comment]
@FrankEssenberger Hi Frank! We've done the steps you asked and found Unauthorized errors in the logs of the on-premise system. We also did successful get requests from the working application and it can be seen in logs also. Could we arrange a call and review logs as there is lots of sensitive information in them so I would avoid attaching them here? 

Kind Regards,
Evgeniy

**Comment by FrankEssenberger** [comment]
We spent some time today to look at the issue and there is still something to clarify.

However it looks as if there is some problem with a generated VDM. So when we tried in the same app a call to the business partner standard API we got a 403 error instead of the 401 with the VDM for the custom service. I have the .edmx file for the service if we need this for investigation. 

Evgeniy will do two things:
- Ask to increase the permission of the test user on ISD so that the business partner call works fine
- Regenerate the VDM once more with the latest version of the generator in the sdk.

If the error still persists I will ask about the credentials for a test user and try to investigate the issue first locally and then from CF.
